File: Implementation/Alphabet Animals.java
// NCPC 2019 (Nordic Collegiate Programming Contest)
// Problem A (Alphabet Animals)

import java.io.*;
import java.util.*;
 
public class Main {
	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
		StringTokenizer st = new StringTokenizer(br.readLine());
		
		String prev = st.nextToken();
		st = new StringTokenizer(br.readLine());
		int n = Integer.parseInt(st.nextToken());
		
		ArrayList<String>[] words = new ArrayList[26];
		for(int i=0; i<26; i++) {
		    words[i] = new ArrayList<>();
		}
		
		for(int i=0; i<n; i++) {
		    st = new StringTokenizer(br.readLine());
		    String w = st.nextToken();
		    words[w.charAt(0) - 'a'].add(w);
		}
		
		String answer = "";
		int start = prev.charAt(prev.length()-1) - 'a';
		for(int i=0; i<words[start].size(); i++) {
		    String current = words[start].get(i);
		    
		    if(answer.length() == 0) {
		        answer = words[start].get(0);
		    }
		    
		    int next = current.charAt(current.length()-1) - 'a';
		    if(words[next].size() == 0 || (words[next].size() == 1 && start == next)) {
		        answer = words[start].get(i) + "!";
		    }
		    
		    if(answer.charAt(answer.length()-1) == '!') break;
		}
		
		if(answer.length() == 0) {
		    bw.write("?\n");
		} else {
		    bw.write(answer);
		}
		bw.flush();
		bw.close();
	}
}

File: Implementation/Alphabet Animals.py
# NCPC 2019 (Nordic Collegiate Programming Contest)
# Problem A (Alphabet Animals)

import sys
 
prev = sys.stdin.readline().strip()
n = int(sys.stdin.readline())
words = [[] for _ in range(26)]
 
for i in range(n):
    tmp = sys.stdin.readline().strip()
    words[ord(tmp[0]) - ord('a')].append(tmp)
 
select = ""
start = ord(prev[len(prev) - 1]) - ord('a')
can_eliminate = False
for i in range(len(words[start])):
    if select == "": select = words[start][0]
    
    length = len(words[start][i])
    next_start = words[start][i][length-1]
    
    if next_start != words[start][i][0]:
        if len(words[ord(next_start) - ord('a')]) == 0:
            select = words[start][i]
            can_eliminate = True
    
    else:
        if len(words[ord(next_start) - ord('a')]) == 1:
            select = words[start][i]
            can_eliminate = True
 
    if can_eliminate == True: break
    
if select == "": print("?")
elif select != "" and can_eliminate == False: print(select)
else: print(select + "!")
